src/wp-includes/class-wp-textdomain-registry.php

```diff
@@ -0,0 +1,184 @@
+	 *
+	 * @since 6.1.0
+	 *
+	 * @param string $domain Text domain.
+	 * @return string|false Current MO file path or false if there is none available.
+	 */
+	public function get_current( $domain ) {
+		if ( isset( $this->current[ $domain ] ) ) {
+			return $this->current[ $domain ];
+		}
+
+		return false;
+	}
+
+	/**
+	 * Sets the MO file path for a specific domain and locale.
+	 *
+	 * Also sets the 'current' property for direct access
+	 * to the path for the current (most recent) locale.
+	 *
+	 * @since 6.1.0
+	 *
+	 * @param string       $domain Text domain.
+	 * @param string       $locale Locale.
+	 * @param string|false $path   Language directory path or false if there is none available.
+	 */
+	public function set( $domain, $locale, $path ) {
+		$this->all[ $domain ][ $locale ] = $path ? trailingslashit( $path ) : false;
+		$this->current[ $domain ]        = $this->all[ $domain ][ $locale ];
+	}
+
+	/**
+	 * Resets the registry state.
+	 *
+	 * @since 6.1.0
+	 */
+	public function reset() {
+		$this->cached_mo_files = null;
+		$this->all             = array();
+		$this->current         = array();
+	}
+
+	/**
+	 * Gets the path to a translation file in the languages directory for the current locale.
+	 *
+	 * @since 6.1.0
+	 *
+	 * @param string $domain Text domain.
+	 * @param string $locale Locale.
+	 * @return string|false MO file path or false if there is none available.
+	 */
+	private function get_path_from_lang_dir( $domain, $locale ) {
+		if ( null === $this->cached_mo_files ) {
+			$this->set_cached_mo_files();
+		}
+
+		$mofile = "{$domain}-{$locale}.mo";
+
+		$path = WP_LANG_DIR . '/plugins/' . $mofile;
+
+		if ( in_array( $path, $this->cached_mo_files, true ) ) {
+			$path = WP_LANG_DIR . '/plugins/';
+			$this->set( $domain, $locale, $path );
+
+			return $path;
+		}
+
+		$path = WP_LANG_DIR . '/themes/' . $mofile;
+		if ( in_array( $path, $this->cached_mo_files, true ) ) {
+			$path = WP_LANG_DIR . '/themes/';
+			$this->set( $domain, $locale, $path );
+
+			return $path;
+		}
+
+		// If no path is found for the given locale, check if an entry for the default
+		// en_US locale exists. This is the case when e.g. using load_plugin_textdomain
+		// with a custom path.
+		if ( 'en_US' !== $locale && isset( $this->all[ $domain ]['en_US'] ) ) {
+			$this->set( $domain, $locale, $this->all[ $domain ]['en_US'] );
+			return $this->all[ $domain ]['en_US'];
+		}
+
+		$this->set( $domain, $locale, false );
+
+		return false;
+	}
+
+	/**
+	 * Reads and caches all available MO files from the plugins and themes language directories.
+	 *
+	 * @since 6.1.0
+	 */
+	protected function set_cached_mo_files() {
+		$this->cached_mo_files = array();
+
+		$locations = array(
+			WP_LANG_DIR . '/plugins',
+			WP_LANG_DIR . '/themes',
+		);
+
+		foreach ( $locations as $location ) {
+			$mo_files = glob( $location . '/*.mo' );
+
+			if ( $mo_files ) {
+				$this->cached_mo_files = array_merge( $this->cached_mo_files, $mo_files );
```

Comment:
You think there's some opportunity to remove the `array_merge()` call in the loop?

Reply:
`$this->cached_mo_files = array( ...$this->cached_mo_files, ...$mo_files );` perhaps? Not sure if that's supported on PHP 5.6.

FWIW this function should only be running once, and this part is just copied from `_get_path_to_translation_from_lang_dir`

Reply:
>  Array unpacking within array declarations using the spread operator is not supported in PHP 7.3 or earlier.

So... yeah I don't see a simpler way right now.